// @primer/components/lib/Pagination/Pagination.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _sx = _interopRequireDefault(require("../sx"));

var _constants = require("../constants");

var _theme = _interopRequireDefault(require("../theme"));

var _Box = _interopRequireDefault(require("../Box"));

var _model = require("./model");

var Page = _styledComponents["default"].a.withConfig({
  displayName: "Pagination__Page",
  componentId: "sc-58gpf2-0"
})(["position:relative;float:left;padding:7px 12px;margin-left:-1px;font-size:", ";font-style:normal;font-weight:", ";color:", ";white-space:nowrap;vertical-align:middle;cursor:pointer;user-select:none;background:", ";border:", " solid ", ";text-decoration:none;&:first-child{margin-left:0;border-top-left-radius:", ";border-bottom-left-radius:", ";}&:last-child{border-top-right-radius:", ";border-bottom-right-radius:", ";}", " ", " ", " ", ";"], (0, _constants.get)('pagination.fontSize'), (0, _constants.get)('pagination.fontWeight'), (0, _constants.get)('pagination.colors.normal.fg'), (0, _constants.get)('pagination.colors.normal.bg'), (0, _constants.get)('borderWidths.1'), (0, _constants.get)('pagination.colors.normal.border'), (0, _constants.get)('pagination.borderRadius'), (0, _constants.get)('pagination.borderRadius'), (0, _constants.get)('pagination.borderRadius'), (0, _constants.get)('pagination.borderRadius'), function (props) {
  return !props.selected && !props.disabled && (0, _styledComponents.css)(["&:hover,&:focus{z-index:2;color:", ";background-color:", ";border-color:", ";}"], (0, _constants.get)('pagination.colors.hover.fg'), (0, _constants.get)('pagination.colors.hover.bg'), (0, _constants.get)('pagination.colors.hover.border'));
}, function (props) {
  return props.selected && (0, _styledComponents.css)(["z-index:3;color:", ";background-color:", ";border-color:", ";"], (0, _constants.get)('pagination.colors.selected.fg'), (0, _constants.get)('pagination.colors.selected.bg'), (0, _constants.get)('pagination.colors.selected.border'));
}, function (props) {
  return props.disabled && (0, _styledComponents.css)(["color:", ";cursor:default;background-color:", ";border-color:", ";"], (0, _constants.get)('pagination.colors.disabled.fg'), (0, _constants.get)('pagination.colors.disabled.bg'), (0, _constants.get)('pagination.colors.disabled.border'));
}, _constants.COMMON);

function usePaginationPages(_ref) {
  var theme = _ref.theme,
      pageCount = _ref.pageCount,
      currentPage = _ref.currentPage,
      onPageChange = _ref.onPageChange,
      hrefBuilder = _ref.hrefBuilder,
      marginPageCount = _ref.marginPageCount,
      showPages = _ref.showPages,
      surroundingPageCount = _ref.surroundingPageCount;

  var pageChange = _react["default"].useCallback(function (n) {
    return function (e) {
      return onPageChange(e, n);
    };
  }, [onPageChange]);

  var model = _react["default"].useMemo(function () {
    return (0, _model.buildPaginationModel)(pageCount, currentPage, showPages, marginPageCount, surroundingPageCount);
  }, [pageCount, currentPage, showPages, marginPageCount, surroundingPageCount]);

  var children = _react["default"].useMemo(function () {
    return model.map(function (page) {
      var _buildComponentData = (0, _model.buildComponentData)(page, hrefBuilder, pageChange(page.num)),
          props = _buildComponentData.props,
          key = _buildComponentData.key,
          content = _buildComponentData.content;

      return /*#__PURE__*/_react["default"].createElement(Page, (0, _extends2["default"])({}, props, {
        key: key,
        theme: theme
      }), content);
    });
  }, [model, hrefBuilder, pageChange, theme]);

  return children;
}

var PaginationContainer = _styledComponents["default"].nav.withConfig({
  displayName: "Pagination__PaginationContainer",
  componentId: "sc-58gpf2-1"
})(["margin-top:20px;margin-bottom:15px;text-align:center;", ";"], _sx["default"]);

function Pagination(_ref2) {
  var theme = _ref2.theme,
      pageCount = _ref2.pageCount,
      currentPage = _ref2.currentPage,
      onPageChange = _ref2.onPageChange,
      hrefBuilder = _ref2.hrefBuilder,
      marginPageCount = _ref2.marginPageCount,
      showPages = _ref2.showPages,
      surroundingPageCount = _ref2.surroundingPageCount,
      rest = (0, _objectWithoutProperties2["default"])(_ref2, ["theme", "pageCount", "currentPage", "onPageChange", "hrefBuilder", "marginPageCount", "showPages", "surroundingPageCount"]);
  var pageElements = usePaginationPages({
    theme: theme,
    pageCount: pageCount,
    currentPage: currentPage,
    onPageChange: onPageChange,
    hrefBuilder: hrefBuilder,
    marginPageCount: marginPageCount,
    showPages: showPages,
    surroundingPageCount: surroundingPageCount
  });
  return /*#__PURE__*/_react["default"].createElement(PaginationContainer, (0, _extends2["default"])({
    "aria-label": "Pagination"
  }, rest, {
    theme: theme
  }), /*#__PURE__*/_react["default"].createElement(_Box["default"], {
    display: "inline-block",
    theme: theme
  }, pageElements));
}

Pagination.displayName = "Pagination";

function defaultHrefBuilder(pageNum) {
  return "#".concat(pageNum);
}

function noop() {}

Pagination.propTypes = (0, _objectSpread2["default"])({
  currentPage: _propTypes["default"].number.isRequired,
  hrefBuilder: _propTypes["default"].func,
  marginPageCount: _propTypes["default"].number,
  onPageChange: _propTypes["default"].func,
  pageCount: _propTypes["default"].number.isRequired,
  showPages: _propTypes["default"].bool,
  surroundingPageCount: _propTypes["default"].number
}, _constants.COMMON.propTypes, {}, _sx["default"].propTypes);
Pagination.defaultProps = {
  hrefBuilder: defaultHrefBuilder,
  marginPageCount: 1,
  onPageChange: noop,
  showPages: true,
  surroundingPageCount: 2,
  theme: _theme["default"]
};
var _default = Pagination;
exports["default"] = _default;
